sm3/sm3_types.cpp

```diff
@@ -98,52 +107,85 @@ std::ostream& operator << (std::ostream& os, OpCode op) {
-    case SemanticUsage::eDepth:        return os << "Depth";
-    case SemanticUsage::eSample:       return os << "Sample";
+    case SemanticUsage::ePosition:     return os << "position";
+    case SemanticUsage::eBlendWeight:  return os << "weight";
+    case SemanticUsage::eBlendIndices: return os << "blend";
+    case SemanticUsage::eNormal:       return os << "normal";
+    case SemanticUsage::ePointSize:    return os << "psize";
+    case SemanticUsage::eTexCoord:     return os << "texcoord";
+    case SemanticUsage::eTangent:      return os << "tangent";
+    case SemanticUsage::eBinormal:     return os << "binormal";
+    case SemanticUsage::eTessFactor:   return os << "tessfactor";
+    case SemanticUsage::ePositionT:    return os << "positiont";
+    case SemanticUsage::eColor:        return os << "color";
+    case SemanticUsage::eFog:          return os << "fog";
+    case SemanticUsage::eDepth:        return os << "depth";
+    case SemanticUsage::eSample:       return os << "sample";
   }
 
-  return os << "Usage(" << uint32_t(usage) << ")";
+  return os << "SemanticUsage(" << uint32_t(usage) << ")";
 }
 
 std::ostream& operator << (std::ostream& os, TextureType textureType) {
   switch (textureType) {
-    case TextureType::eTexture2D:   return os << "Texture2D";
-    case TextureType::eTextureCube: return os << "TextureCube";
-    case TextureType::eTexture3D:   return os << "Texture3D";
+    case TextureType::eTexture2D:   return os << "2d";
+    case TextureType::eTextureCube: return os << "cube";
+    case TextureType::eTexture3D:   return os << "3d";
   }
 
   return os << "TextureType(" << uint32_t(textureType) << ")";
 }
 
 std::ostream& operator << (std::ostream& os, RasterizerOutIndex outIndex) {
   switch (outIndex) {
-    case RasterizerOutIndex::eRasterOutPosition:  return os << "RasterizerOutPosition";
-    case RasterizerOutIndex::eRasterOutFog:       return os << "RasterizerOutFog";
-    case RasterizerOutIndex::eRasterOutPointSize: return os << "RasterizerOutPointSize";
+    case RasterizerOutIndex::eRasterOutPosition:  return os << "Pos";
+    case RasterizerOutIndex::eRasterOutFog:       return os << "Fog";
+    case RasterizerOutIndex::eRasterOutPointSize: return os << "PointSize";
   }
 
-  return os << "RasterizerOutIndex(" << uint32_t(outIndex) << ")";
+  return os << "RasterizerOutValue(" << uint32_t(outIndex) << ")";
 }
 
 std::ostream& operator << (std::ostream& os, MiscTypeIndex miscTypeIndex) {
   switch (miscTypeIndex) {
-    case MiscTypeIndex::eMiscTypePosition: return os << "MiscTypePosition";
-    case MiscTypeIndex::eMiscTypeFace:     return os << "MiscTypeFace";
+    case MiscTypeIndex::eMiscTypePosition: return os << "Pos";
+    case MiscTypeIndex::eMiscTypeFace:     return os << "Face";
   }
 
   return os << "MiscTypeIndex(" << uint32_t(miscTypeIndex) << ")";
 }
 
+std::ostream& writeToStream(std::ostream& os, RegisterType registerType, ShaderType shaderType, uint32_t shaderVersionMajor) {
```

Comment:
This could also just be an `operator <<` that takes a struct, think that's cleaner than adding another global function with an overly generic name where nobody is going to know what this is even supposed to do a week from now.